obs-studio: fix NVENC

VeilSilence opened the pull request:
After updating obs-studio from 30.2.3 to 31.0.1, NVENC options stopped working. This PR fixes the issue and restores proper NVENC support.
Closes https://github.com/NixOS/nixpkgs/issues/382666

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

VeilSilence committed:
obs-studio: fix NVENC

fpletz commented:
Awesome! Thank you! I currently don't have access to my machine with a nvidia GPU so I couldn't address this yet. Unfortunately that also means I can't review this this right now. If anybody could test this I will gladly merge.

One thing: The commit title should be same as you have correctly selected for this PR. We can also fix this by squashing though.

VeilSilence commented:
> Awesome! Thank you! I currently don't have access to my machine with a nvidia GPU so I couldn't address this yet. Unfortunately that also means I can't review this this right now. If anybody could test this I will gladly merge.
> 
> One thing: The commit title should be same as you have correctly selected for this PR. We can also fix this by squashing though.

I'm thinking about if we should make `autoAddDriverRunpath` Linux only
About title sorry, I'm new to all of this so

VeilSilence commented:
Updated pull request. 
Added `lib.optionals cudaSupport`, because that how most other apps do like sunshine, ollama etc.

fxzzi commented:
can confirm this fixes the issue, I added your fork as a flake input and enabled cuda support by setting it like so in home-manager:
```nix
programs.obs-studio.package = inputs.nixpkgs-obs-nvenc.legacyPackages.${pkgs.system}.obs-studio.override {
    cudaSupport = true;
};
```

nvenc is available again :)

VeilSilence commented:
Updated again. Removed redundant `lib` in `optional`. 

> can confirm this fixes the issue, I added your fork as a flake input and enabled cuda support by setting it like so in home-manager:
> 
> ```nix
> programs.obs-studio.package = inputs.nixpkgs-obs-nvenc.legacyPackages.${pkgs.system}.obs-studio.override {
>     cudaSupport = true;
> };
> ```
> 
> nvenc is available again :)

There is also `nixpkgs.config.cudaSupport = true;` just in case.

fxzzi commented:
> There is also `nixpkgs.config.cudaSupport = true;` just in case.

seems like this doesn't make the package build with cuda for me. needed the override for it to build correctly

VeilSilence commented:
> > There is also `nixpkgs.config.cudaSupport = true;` just in case.
> 
> seems like this doesn't make the package build with cuda for me. needed the override for it to build correctly

Hmm, I'm using my fork as `nixpkgs.url` and it's working for me.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/hardware-encoding-in-obs-studio-with-nvidia-rtx-2070-tu106/63638/3
